src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyDispatcher.cs

```diff
@@ -0,0 +1,163 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Rendering;
+
+// When Blazor is deployed with multi-threaded runtime, WebAssemblyDispatcher will help to dispatch all Blazor JS interop calls to the main thread.
+// This is necessary because all JS objects have thread affinity. They are only available on the thread (WebWorker) which created them.
+// Also DOM is only available on the main (browser) thread.
+// Because all of the Dispatcher.InvokeAsync methods return Task, we don't need to propagate errors via OnUnhandledException handler
+internal sealed class WebAssemblyDispatcher : Dispatcher
+{
+    public static readonly Dispatcher Instance = new WebAssemblyDispatcher();
+    private readonly SynchronizationContext _mainSyncContext;
+
+    private WebAssemblyDispatcher()
+    {
+        Debug.Assert(Environment.CurrentManagedThreadId == 1); // WebAssemblyDispatcher should be initialized on the main thread.
+
+        // capture the JSSynchronizationContext from the main thread, which runtime already installed.
+        _mainSyncContext = SynchronizationContext.Current!;
```

Comment:
should we ! this instead of asserting that it's not null? someone could have switched it

Reply:
Relevant discussion was on previous attempt here https://github.com/dotnet/aspnetcore/pull/48991#discussion_r1258087940

My understanding is that if people install another `SynchronizationContext` than `JSSynchronizationContext` it's up to them to make it work.

I could stop asserting the `JSSynchronizationContext` by name to allow it. What do you think ?

`JSSynchronizationContext` is not public API so I could not re-install it here.

Reply:
I simplified the code and added comments.